lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx

```diff
@@ -0,0 +1,542 @@
+  const NumericalMathFunction model(problem.getObjective());
+  NumericalSample inputSample(krigingResult_.getInputSample());
+  NumericalSample outputSample(model(inputSample));
+  UnsignedInteger size = inputSample.getSize();
+  NumericalPoint noise(size);
+  Bool hasNoise = model.getOutputDimension() == 2;
+  if (hasNoise)
+  {
+    // use noise model to optimize AEI else fallback to objective 2nd marginal
+    if (noiseModel_.getOutputDimension() != 1)
+      noiseModel_ = model.getMarginal(1);
+
+    NumericalSample noiseSample(outputSample.getMarginal(1));
+    outputSample = outputSample.getMarginal(0);
+    for (UnsignedInteger i = 0; i < size; ++ i)
+    {
+      noise[i] = noiseSample[i][0];
+      if (!(noise[i] >= 0.0)) throw InvalidArgumentException(HERE) << "Noise model must be positive";
+    }
+  }
+  UnsignedInteger iterationNumber = 0;
+  Bool convergence = false;
+
+  // select the best feasible point
+  NumericalPoint optimizer;
+  NumericalScalar optimalValue = problem.isMinimization() ? SpecFunc::MaxNumericalScalar : -SpecFunc::MaxNumericalScalar;
+  NumericalPoint optimizerPrev; // previous optimizer
+  NumericalScalar optimalValuePrev = optimalValue;// previous optimal value
+  for (UnsignedInteger index = 0; index < size; ++ index)
+  {
+    if (!problem.hasBounds() || (problem.hasBounds() && problem.getBounds().contains(inputSample[index])))
+      if ((problem.isMinimization() && (outputSample[index][0] < optimalValue))
+      || (!problem.isMinimization() && (outputSample[index][0] > optimalValue)))
+      {
+        optimizerPrev = optimizer;
+        optimalValuePrev = optimalValue;
+
+        optimizer = inputSample[index];
+        optimalValue = outputSample[index][0];
+      }
+  }
+
+  LOGINFO(OSS() << "Optimum so far x=" << optimizer << " f(x)=" << optimalValue);
+
+  // we need the second best to compute convergence criteria
+  if (optimizerPrev.getDimension() == 0)
+  {
+    // then the optimum was the first
+    for (UnsignedInteger index = 1; index < size; ++ index)
+    {
+      if (!problem.hasBounds() || (problem.hasBounds() && problem.getBounds().contains(inputSample[index])))
+        if ((problem.isMinimization() && (outputSample[index][0] < optimalValuePrev))
+        || (!problem.isMinimization() && (outputSample[index][0] > optimalValuePrev)))
+        {
+          optimizerPrev = inputSample[index];
+          optimalValuePrev = outputSample[index][0];
+        }
+    }
+  }
+
+  // compute minimum distance
+  NumericalPoint minimumDistance(dimension, SpecFunc::MaxNumericalScalar);
+  if (!hasNoise)
+  {
+    for (UnsignedInteger i1 = 0; i1 < size; ++ i1)
+    {
+      for (UnsignedInteger i2 = 0; i2 < i1; ++ i2)
+      {
+        for (UnsignedInteger j = 0; j < dimension; ++ j)
+        {
+          NumericalScalar distance = std::abs(inputSample[i1][j] - inputSample[i2][j]);
+          if (distance < minimumDistance[j])
+          {
+            minimumDistance[j] = distance;
+          }
```

Comment:
The cost is dn(n-1)/2. If n is large, it can be time-consuming. If you sort each marginal sample (cost: d*n*log(n)) then you can find the minimum distance per component in an additional d*n step, for a total cost dominated by the first step.